pkg/machines/helpers.js

```diff
@@ -375,3 +377,164 @@ export function timeoutedPromise(promise, delay, afterTimeoutHandler, afterTimeo
+                d.capability.lun._value == unit &&
+                d.capability.target._value == target)
+                return true;
+        });
+        break;
+    }
+    case "scsi_host": {
+        // TODO add scsi_host
+        nodeDev = undefined;
+        break;
+    }
+    case "mdev": {
+        const uuid = hostdev.source.address.uuid;
+        nodeDev = nodeDevices.find(d => {
+            if (d.path &&
+                d.path._value.contains(uuid))
+                return true;
+        });
+        break;
+    }
+    }
+    return nodeDev;
+}
+
+export function getBootOrderDevices(vm) {
+    let devices = [];
+    let diskCount = 0;
+    let ifaceCount = 0;
+
+    // Some disks and interfaces may have boot order in os->boot (legacy)
+    if (vm.osBoot) {
+        for (let i = 0; i < vm.osBoot.length; i++) {
+            const boot = vm.osBoot[i];
+
+            if (boot.type === "disk" | boot.type === "cdrom" | boot.type === "fd") {
+                devices.push({
+                    device: Object.values(vm.disks)[diskCount++],
+                    bootOrder: i + 1, // bootOrder begins at 1
+                    type: "disk"
+                });
+            } else if (boot.type === "network") {
+                devices.push({
+                    device: Object.values(vm.interfaces)[ifaceCount++],
+                    bootOrder: i + 1, // bootOrder begins at 1
+                    type: "network"
+                });
+            }
+        }
+    }
+
+    const disks = Object.values(vm.disks);
+    // we ignore the disks which have been loaded from os->boot, and load the rest of disks
+    for (let i = diskCount; i < disks.length; i++) {
+        const disk = disks[i];
+
+        devices.push({
+            device: disk,
+            bootOrder: disk.bootOrder,
+            type: "disk"
+        });
+    }
+
+    const ifaces = Object.values(vm.interfaces);
+    // we ignore the interfaces which have been loaded from os->boot, and load the rest of interfaces
+    for (let i = ifaceCount; i < ifaces.length; i++) {
+        const iface = ifaces[i];
+
+        devices.push({
+            device: iface,
+            bootOrder: iface.bootOrder,
+            type: "network"
+        });
+    }
+
+    // redirected devices cannot have os->boot boot order
+    Object.values(vm.redirectedDevices)
+            .forEach(redirdev => {
+                devices.push({
+                    device: redirdev,
+                    bootOrder: redirdev.bootOrder,
+                    type: "redirdev"
+                });
+            });
+
+    // host devices cannot have os->boot boot order
+    Object.values(vm.hostDevices)
+            .forEach(hostdev => {
+                devices.push({
+                    device: hostdev,
+                    bootOrder: hostdev.bootOrder,
+                    type: "hostdev"
+                });
+            });
+
+    return devices;
+}
+
+export function getSortedBootOrderDevices(vm) {
```

Comment:
Please explain the logic behind this sorting. Either inline with comments either above the function definition.